### PR TITLE
Implemented code signing parameter for Azure Trusted Signing

### PR DESCRIPTION
**Pull request by madorjan**
Implemented a parameter for trusted signing as suggested here

https://github.com/velopack/velopack/issues/142

To discuss:
- How to reference Azure.CodeSigning.Dlib? Using https://www.nuget.org/packages/Microsoft.Trusted.Signing.Client?
- Timestamp server + algorithms configurable or hard coded?

### CODE REVIEW

**Review by Keboo — COMMENTED** [review]
Thank you for doing this! It is great work.

For loading up the Azure.CodeSigning.Dlib I think going through the NuGet package is probably simplest. We likely need to make an x86/x64 check to make sure we load up the right version of the assembly (though the props file in the NuGet package is just hardcoded to the x64 path). Let me know if you run into any issues with this.

File: src/vpk/Velopack.Packaging.Windows/Commands/WindowsPackCommandRunner.cs [hunk 1]
```diff
@@ -261,7 +262,12 @@ private void SignFilesImpl(WindowsSigningOptions options, Action<int> progress,
         // signtool.exe does not work if we're not on windows.
         if (!VelopackRuntimeInfo.IsWindows) return;
 
-        if (!string.IsNullOrEmpty(signParams)) {
+        if(!string.IsNullOrEmpty(trustedSignMetadataPath)) {
+            Log.Info($"Use Azure Trusted Signing service for code signing. Metadata file path: {trustedSignMetadataPath}");
+            signParams = $"/fd SHA256 /tr \"http://timestamp.acs.microsoft.com\" /v /debug /td SHA256 /dlib \"{HelperFile.AzTrustedSigningDlibPath}\" /dmdf \"{trustedSignMetadataPath}\"";
```
[the remark this answers is not in the document]
Comment (caesay) [reply src/vpk/Velopack.Packaging.Windows/Commands/WindowsPackCommandRunner.cs:267]: Just as a second data point, I don't mind personally these being hardcoded (and then seeing if anyone complains and asks it to be configurable)

**Comment by caesay** [comment]
> For loading up the Azure.CodeSigning.Dlib I think going through the NuGet package is probably simplest. 

We do have code in `NugetDownloader.cs` which is capable of downloading a NuGet to disk already. That could be moved from Vpk project to Packaging project and used on-demand to download the Dlib package when the az trusted signing is used. We should probably just download it once / the first time it's used and cache it somewhere (maybe beside vendor folder?)

**Comment by madorjan** [comment]
Thanks for continueing this pull request. Unfortunately, I haven't had time recently ...

Any idea how to fix the signtool versioning issue in the long-term? Maybe, we have to download the latest "Microsoft.Windows.SDK.BuildTools" via the NugetDownloader too, to ensure that we have always the latest versions of signtool.exe and dlib.dll?

**Comment by Keboo** [comment]
No problem, I appreciate you getting the ball rolling for this.

For the `SignTool.exe` that is an ongoing discussion. At least at present, the current version that is in vpk meets the minimum requirements. 
There are some ongoing discussions on how to best handle the versioning of these. This is just the first step to get it working. Technically Dlib still has the same versioning problem as it only downloads when the dll is not present. So for persistent systems (like dev boxes) it will only download once and wont update it again. 

I think we will need to provide some mechanism to make it easy for people to "get latest" or "get version" (for both Dlib and signtool).

**Comment by caesay** [comment]
Superseded by #330

**Pull request closed without merging**